Replace FutureMessage with ivalue::Future in RRefContext

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50236 Apply clang-format to rpc cpp files
* #50029 Completely remove FutureMessage type
* #50028 Completely Remove FutureMessage from RPC agents
* #50027 Completely Remove FutureMessage from RPC cpp tests
* #50026 Remove FutureMessage from RPC request callback logic
* #50025 Completely remove FutureMessage from FaultyProcessGroupAgent
* #50024 Remove FutureMessage from sender TensorPipeAgent
* #50023 Remove FutureMessage from sender ProcessGroupAgent
* #50020 Completely remove FutureMessage from distributed autograd
* #50005 Completely remove FutureMessage from RPC TorchScript implementations
* #50004 Completely remove FutureMessage from RRef Implementations
* #49995 Replace FutureMessage with ivalue::Future in RpcAgent retry logic
* **#49960 Replace FutureMessage with ivalue::Future in RRefContext**
* #49927 Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*
* #49906 Let RpcAgent::send() return JitFuture

Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

mrshenli committed:
Replace FutureMessage with ivalue::Future in RRefContext

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -377,7 +377,7 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
   }
 
   // Get the result of the current future.
-  IValue value() {
```
Are IValues immutable? Do they have value or pointer semantics?

My worry is that, if this function effectively returns a new _pointer_ to the same value, then the caller could modify it and thus (in some sense) modify the future even if it only had const access to it.

lw replied:
I see that below constValue is being replaced by value in order to call toCustomClass on it. It seems toCustomClass should be able to operate on const& (there's a specific overload qualified as such). If that is the reason for this change, I'd investigate why toCustomClass doesn't work as intended...

mrshenli replied:
Yep, constValue works, let me update.

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]

mrshenli committed:
Update on "Replace FutureMessage with ivalue::Future in RRefContext"


Differential Revision: [D25730530](https://our.internmc.facebook.com/intern/diff/D25730530)

[ghstack-poisoned]
